x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java

```diff
@@ -217,8 +218,24 @@ public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress, Sock
                 sslParameters.setServerNames(Collections.singletonList(serverName));
                 sslEngine.setSSLParameters(sslParameters);
             }
-            ctx.pipeline().replace(this, "ssl", new SslHandler(sslEngine));
-            super.connect(ctx, remoteAddress, localAddress, promise);
+            final ChannelPromise connectPromise = ctx.newPromise();
+            final SslHandler sslHandler = new SslHandler(sslEngine);
+            ctx.pipeline().replace(this, "ssl", sslHandler);
+            final Future<?> handshakePromise = sslHandler.handshakeFuture();
+            connectPromise.addListener(result -> {
+                if (result.isSuccess() == false) {
+                    promise.tryFailure(result.cause());
+                } else {
+                    handshakePromise.addListener(handshakeResult -> {
+                        if (handshakeResult.isSuccess()) {
+                            promise.setSuccess();
+                        } else {
+                            promise.tryFailure(handshakeResult.cause());
```

Comment:
Went for `tryFailure` here out of paranoia, probably not necessary but didn't want to risk an uncaught exception here.